util/timer/src/timer.rs

```diff
@@ -161,17 +193,37 @@ impl Scheduler {
         result
     }
 
+    fn clear_and_stop(&self) {
+        let mut scheduler = self.inner.lock();
+        scheduler.clear_and_stop();
+        self.condvar.notify_all();
+    }
+
+    fn start_drain(&self) {
+        let mut scheduler = self.inner.lock();
+        scheduler.finished = true;
+        self.condvar.notify_all();
+    }
+
     fn run(&self, worker_queue: &WorkerQueue) {
         let mut scheduler = self.inner.lock();
-        while !scheduler.stop {
+        loop {
             let wait_for = scheduler.handle_timeout(worker_queue);
             match wait_for {
                 Some(timeout) => {
                     self.condvar.wait_for(&mut scheduler, timeout);
                 }
-                None => self.condvar.wait(&mut scheduler),
+                None => {
+                    if scheduler.finished {
```

Comment:
Why do you check finished here? Do we have to wait for timeout even the scheduler finished?